pkg/controller/catalog/datacatalog/datacatalog.go

```diff
@@ -0,0 +1,292 @@
+		if err := m.store.ReadProtobuf(ctx, inputPath, inputs); err != nil {
+			logger.Errorf(ctx, "DataCatalog failed to read inputs %+v, err: %+v", inputPath, err)
+			return err
+		}
+		logger.Debugf(ctx, "DataCatalog read inputs from %v", inputPath)
+	}
+
+	if task.Interface.Outputs != nil && len(task.Interface.Outputs.Variables) != 0 {
+		if err := m.store.ReadProtobuf(ctx, outputPath, outputs); err != nil {
+			logger.Errorf(ctx, "DataCatalog failed to read outputs %+v, err: %+v", outputPath, err)
+			return err
+		}
+		logger.Debugf(ctx, "DataCatalog read outputs from %v", outputPath)
+	}
+
+	datasetID, err := transformer.GenerateDatasetIDForTask(ctx, task)
+	if err != nil {
+		logger.Errorf(ctx, "DataCatalog failed to generate dataset for task %+v, err: %+v", task, err)
+		return err
+	}
+
+	logger.Debugf(ctx, "DataCatalog put into Catalog for DataSet %v", datasetID)
+
+	// Try creating the dataset in case it doesn't exist
+	metadata := &datacatalog.Metadata{
+		KeyMap: map[string]string{
+			taskVersionKey: task.Id.Version,
+			taskExecKey:    execID.NodeExecutionId.NodeId,
+		},
+	}
+	newDataset := &datacatalog.Dataset{
+		Id:       datasetID,
+		Metadata: metadata,
+	}
+
+	_, err = m.client.CreateDataset(ctx, &datacatalog.CreateDatasetRequest{Dataset: newDataset})
+	if err != nil {
+		logger.Debugf(ctx, "Create dataset %v return err %v", datasetID, err)
+
+		if status.Code(err) == codes.AlreadyExists {
+			logger.Debugf(ctx, "Create Dataset for task %v already exists", task.Id)
+		} else {
+			logger.Errorf(ctx, "Unable to create dataset %+v, err: %+v", datasetID, err)
+			return err
+		}
+	}
+
+	// Create the artifact for the execution that belongs in the task
+	artifactDataList := make([]*datacatalog.ArtifactData, 0, len(outputs.Literals))
+	for name, value := range outputs.Literals {
+		artifactData := &datacatalog.ArtifactData{
+			Name:  name,
+			Value: value,
+		}
+		artifactDataList = append(artifactDataList, artifactData)
+	}
+
+	cachedArtifact := &datacatalog.Artifact{
+		Id:       string(uuid.NewUUID()),
+		Dataset:  datasetID,
+		Data:     artifactDataList,
+		Metadata: metadata,
+	}
+
+	createArtifactRequest := &datacatalog.CreateArtifactRequest{Artifact: cachedArtifact}
+	_, err = m.client.CreateArtifact(ctx, createArtifactRequest)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to create Artifact %+v, err: %v", cachedArtifact, err)
+		return err
+	}
+	logger.Debugf(ctx, "Created artifact: %v, with %v outputs from execution %v", cachedArtifact.Id, len(artifactDataList), execID.TaskId.Name)
+
+	// Tag the artifact since it is the cached artifact
+	tagName, err := transformer.GenerateArtifactTagName(ctx, inputs)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to create tag for artifact %+v, err: %+v", cachedArtifact.Id, err)
+		return err
+	}
+	logger.Debugf(ctx, "Created tag: %v, for task: %v", tagName, task.Id)
+
+	// TODO: We should create the artifact + tag in a transaction when the service supports that
```

Comment:
Why?

Reply:
Because it will avoid race conditions, for example if there are two run of the same task that happen in parallel we may not be tagging the artifact we just created in this service call. It's not a huge deal because either of those artifacts will work, just can be unexpected behavior.